### DISCUSSION TITLE
Safe idiomatic Rust API for plugins + embedding

### DISCUSSION THREAD

**Discussion by v3xro** [kcl-lang/kcl#1587] [Ideas]
I would like to use KCL as the configuration language for my application written in Rust. I think the following features would be nice to have:

* serde `Deserialize` support from `ValueRef` -> native rust struct/enum (I have a prototype version that I might share once it's better tested)
* Safe idiomatic interface for plugin functions (that is not extern "C") - I guess a macro could be made that allows you to write functions with types like `fn my_plugin_function(a: String, b: i64) -> Option<String>` that then converts this to `extern "C" my_plugin_function(ctx: *mut kclvm_context_t, ...)`
* Better binding for loading config (that does not go through the protobuf/kclvm interface such as the rust lib currently does)
* Ability to customise builtin modules (e.g. limiting the file/environment read functionality by making it unavailable)

Happy to expand on all of these - am currently trying to integrate KCL in such a way and doubtless these ideas will evolve and I will have some working code to demonstrate how to do this currently.

**Comment by Peefy** [comment]
Hello @v3xro 

I'm glad to see more integration and development of KCL Rust. I think these improvements are very helpful. 👍 Feel free to raise more issues or PRs, and we can further make more changes. ❤️

**Comment by v3xro** [comment]
Do you have a good example of how to register a plugin in Rust? `kclvm_plugin_init` seems to want to do a roundtrip through json strings which I'd rather avoid.

**Comment by Peefy** [comment]
I've opened a PR and give the initial impl and example: https://github.com/kcl-lang/kcl/pull/1599

**Reply by v3xro** [reply]
That look good - certainly better than what I've been trying (`evaluator.add_or_update_global_variable("name_of_fun", ValueRef::func(...))`)
